src/mumble/JackAudio.cpp

```diff
@@ -0,0 +1,618 @@
+				aorJackAudio = new JackAudioOutputRegistrar();
+			} else {
+				airJackAudio = NULL;
+				aorJackAudio = NULL;
+				delete jasys;
+				jasys = NULL;
+			}
+		}
+
+		void destroy() {
+			if (airJackAudio)
+				delete airJackAudio;
+			if (aorJackAudio)
+				delete aorJackAudio;
+			if (jasys) {
+				jasys->close_jack();
+				delete jasys;
+				jasys = NULL;
+			}
+		}
+};
+
+static JackAudioInit jackinit; //unused
+
+JackAudioSystem::JackAudioSystem()
+	: bActive(false)
+	, client(NULL)
+	, in_port(NULL)
+	, output_buffer(NULL)
+	, iBufferSize(0)
+	, bJackIsGood(false)
+	, iSampleRate(0)
+{
+	if (g.s.qsJackAudioOutput.isEmpty()) {
+		iOutPorts = 1;
+	} else {
+		iOutPorts = g.s.qsJackAudioOutput.toInt();
+	}
+	memset((void*)&out_ports, 0, sizeof(out_ports));
+}
+
+JackAudioSystem::~JackAudioSystem() {
+}
+
+void JackAudioSystem::init_jack() {
+
+	output_buffer = NULL;
+	jack_status_t status = static_cast<jack_status_t>(0);
+	int err = 0;
+
+	jack_options_t jack_option = g.s.bJackStartServer ? JackNullOption : JackNoStartServer;
+	client = jack_client_open("mumble", jack_option, &status);
+
+	if (client) {
+		in_port = jack_port_register(client, "input", JACK_DEFAULT_AUDIO_TYPE, JackPortIsInput, 0);
+		if (in_port == NULL) {
+			qWarning("JackAudioSystem: unable to register 'input' port");
+			close_jack();
+			return;
+		}
+
+		bJackIsGood = true;
+		iBufferSize = jack_get_buffer_size(client);
+		iSampleRate = jack_get_sample_rate(client);
+
+		setNumberOfOutPorts(iOutPorts);
+
+		if (bJackIsGood == false) {
+			close_jack();
+			return;
+		}
+
+		err = jack_set_process_callback(client, process_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set process callback - jack_set_process_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		err = jack_set_sample_rate_callback(client, srate_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set sample rate callback - jack_set_sample_rate_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		err = jack_set_buffer_size_callback(client, buffer_size_callback, this);
+		if (err != 0) {
+			qWarning("JackAudioSystem: unable to set buffer size callback - jack_set_buffer_size_callback() returned %i", err);
+			close_jack();
+			return;
+		}
+
+		jack_on_shutdown(client, shutdown_callback, this);
+
+		// If we made it this far, then everything is okay
+		qhInput.insert(QString(), tr("Hardware Ports"));
+		qhOutput.insert(QString::number(1), tr("Mono"));
+		qhOutput.insert(QString::number(2), tr("Stereo"));
+		bJackIsGood = true;
+
+	} else {
+		QStringList errors = jackStatusToStringList(status);
+		qWarning("JackAudioSystem: unable to open jack client due to %i errors:", errors.count());
+		for (int i = 0; i < errors.count(); ++i) {
+			qWarning("JackAudioSystem:  %s", qPrintable(errors.at(i)));
+		}
+		bJackIsGood = false;
+		client = NULL;
+	}
+}
+
+void JackAudioSystem::close_jack() {
+
```

Comment:
Extra space.

Reply:
Sorry, I don't really understand this one.
Do you want me to make an extra space? where? or is it about the extra new empty line?
or is there an extra space (which I do not seem to find) which I should remove?

Reply:
It's about the extra empty line (231).

Reply:
Done